FeatureFlags: manage feature flags outside of settings.Cfg

ryantxu opened the pull request:
#### Goals:
* single/standard way to define a feature flag
* clear way to see what flags exist and basic information about them
* add more constraints for some flags.  IE: require running in dev mode, or with enterprise codebase
* easily trace where feature toggles are used (ie IDE "show references")

#### This PR:
1. creates a list of flags in the backend with additional metadata (name, description, state, etc)
2. generates typescript code that matches the flags
3. moves all featureToggle functions out of `setting.go` and into a `FeatureToggle` class
4. Only sends the enabled flags to the frontend.  Updates the types so we use `flag?: boolean` rather than force a complete object 
5. Makes sure flags that require enterprise license are enabled in the license

#### FUTURE: #43941
1. watch `features.yaml` for changes
2. compile real expressions (currently only "true" works) based on environment and variables
3. show simple UI for all flag states

#### FUTURE future
1. standardize the flag id naming conventions.  require camel|snake case?  avoid dashes?
2. unify various cfg 'enabled' settings as feature flags?
3. Well defined feature flag lifecycle, etc

ryantxu committed:
more feature flag options

ryantxu committed:
more feature flag options

ryantxu committed:
fix tests

ryantxu committed:
generate from tests

ryantxu committed:
generate from tests

ryantxu committed:
generate from tests

ryantxu committed:
add api

ryantxu commented on packages/grafana-data/src/types/config.ts:
```diff
@@ -35,26 +36,6 @@ export enum GrafanaEdition {
- *
- * @public
- */
-export interface FeatureToggles {
```
@mckn -- any thoughts on why/if this needs to bee in `@grafana/data`?  It seems like most of the config stuff really belongs in runtime

ryantxu committed:
merge main

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
use optional flags

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
WithFeatures cleanup

ryantxu committed:
only send enabled flags

ryantxu committed:
feature toggles

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
expose metadata to frontend

ryantxu committed:
expose metadata to frontend

ryantxu committed:
show notices

ryantxu committed:
add link from settings

bergquist commented on pkg/setting/setting_feature_toggles.go:
```diff
@@ -10,65 +13,188 @@ import (
+	// Special flags
+	RequiresDevMode    bool `json:"requiresDevMode,omitempty"`    // can not be enabled in production
+	RequiresEnterprise bool `json:"requiresEnterprise,omitempty"` // requires enterprise executable
+	ModifiesDatabase   bool `json:"modifiesDatabase,omitempty"`   // the feature will modify the SQL tables
```
I think this should be part of the state instead. Once the feature is mature enough to be tested in production we have to avoid breaking changes to the schema.

ryantxu replied:
ie: state=alpha-DB?

FYI: I actually started this effort when starting a feature that adds a SQL table, and was looking for a good way to shout that.  My assumption is that if it got merged to main it is reasonable if future "migration" actually drops all data first :grimacing:

bergquist commented on pkg/setting/setting_feature_toggles.go:
```diff
@@ -10,65 +13,188 @@ import (
+	Id          string             `json:"id"`   // the key listed in grafana.ini
+	Name        string             `json:"name"` // user
+	Description string             `json:"description"`
+	AliasIds    []string           `json:"aliasIds,omitempty"` // ids that will also work (support spelling changes and product renames)
```
Is this really needed? I'm worried that it makes things come complicated. Accidentally enabling/disabling feature is extremely annoying :)

ryantxu replied:
AliasIds?  -- not necessary now now.  It was copied from the "registry" structure we use everywhere in the frontend.  This essentially allows us to fix/clean/rename things without breaking existing configs.

But agree we can leave it off

bergquist commented on pkg/setting/setting_feature_toggles.go:
```diff
@@ -10,65 +13,188 @@ import (
+	RequiresDevMode    bool `json:"requiresDevMode,omitempty"`    // can not be enabled in production
+	RequiresEnterprise bool `json:"requiresEnterprise,omitempty"` // requires enterprise executable
+	ModifiesDatabase   bool `json:"modifiesDatabase,omitempty"`   // the feature will modify the SQL tables
+	FrontendFlag       bool `json:"frontend,omitempty"`           // the change is only seen in the frontend (can change without restart!)
```
Reloading feature without restarting should be supported by both frontend and backend. FGAC is a good example of a feature that could enable at runtime.

ryantxu replied:
:+1:

bergquist reviewed (commented):
:smile: 

Glad to see more thoughts on Toggles! 

During the hackaton we implemented (but didn't get far due to incidents/on-call) https://github.com/grafana/grafana/tree/experimentals/feature-toggle-by-headers which introduces a new framework rather then building on top on what we have. The main reason for that is to make the distinction between a feature being in development mode (tested locally, database migrations) and ready for production/staging more clear. The new "experiments" can MUST always be possible to reload at runtime. But this could also be expressed in a state which you introduce in this PR. 

Enabling toggles without restarting the process is *extremely* valuable when operating Grafana at scale and that's why I favor a new config file(s) so we can mount the same file for multiple instances. 

An interesting idea about having states per toggles would be to allow admins to enable features above a certain state from within Grafana itself.

ryantxu commented on pkg/setting/setting_feature_toggles.go:
```diff
@@ -10,65 +13,188 @@ import (
+
+	// Special flags
+	RequiresDevMode    bool `json:"requiresDevMode,omitempty"`    // can not be enabled in production
+	RequiresEnterprise bool `json:"requiresEnterprise,omitempty"` // requires enterprise executable
```
yes -- though it is also a question of who the metadata is for.  IE unclear to me if I set "accesscontrol" feature flag if that has any affect in OSS or not

Again -- my motivation for this was to try to understand what flags we currently have a basic understanding what they do

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
move to package

ryantxu committed:
update function signatures

ryantxu committed:
generate accessors

ryantxu committed:
generate accessors

ryantxu committed:
merge main

ryantxu committed:
move main items to settings

ryantxu committed:
everything back in featuremgmt

ryantxu committed:
everything back in featuremgmt

ryantxu committed:
remove old file

ryantxu committed:
sync names

ryantxu committed:
merge enterprise features with feature flags

ryantxu commented on pkg/services/featuremgmt/manager.go:
```diff
@@ -0,0 +1,236 @@
+	}
+
+	// TODO: CEL - expression
+	return ff.Expression == "true"
```
@bergquist -- TODO... right now this only supports the "true" expression!

ryantxu commented on pkg/services/featuremgmt/manager.go:
```diff
@@ -0,0 +1,236 @@
+		return false
+	}
+
+	if ff.RequiresLicense && !fm.licensing.FeatureEnabled(ff.Name) {
```
@DanCech -- this will make sure any features that require a license are checked with the license provider

ryantxu commented on packages/grafana-runtime/src/utils/licensing.ts:
```diff
@@ -1,6 +1,12 @@
-export const featureEnabled = (feature: string): boolean => {
-  const { enabledFeatures } = config.licenseInfo;
-  return enabledFeatures && enabledFeatures[feature];
+export const featureEnabled = (feature: boolean | undefined | keyof FeatureToggles): boolean => {
```
@DanCech -- this now uses the same `featureToggle` bucket, but note that the license is checked on the backend before anythign can be enabled

ryantxu commented on pkg/models/licensing.go:
```diff
@@ -14,8 +14,6 @@ type Licensing interface {
 
 	StateInfo() string
 
-	EnabledFeatures() map[string]bool
```
@DanCech -- I removed the enumeration from the main interface.  Is the map really necessary?

ryantxu replied:
@DanCech -- do licenses ever set things to *false*?

DanCech replied:
removing it is fine I think but we'll need to verify in enterprise, the map was intended to just be a list of true values, no false values.

ryantxu committed:
update settings

ryantxu commented on pkg/setting/setting.go:
```diff
@@ -341,8 +341,10 @@ type Cfg struct {
-	FeatureToggles       map[string]bool
+	// Check if a feature toggle is enabled
+	// @deprecated
+	IsFeatureToggleEnabled func(key string) bool // filled in dynamically
```
The circular dependencies make this pretty hard to get rid of :grimacing: 

Most places that used the map reference can be replaced with a refernce to `featuremgmt.FeatureToggles`, but a few places can not.

In particular the `settings.Provider` can not use featuremgmt without a circular ref

ryantxu committed:
more lint

ryantxu committed:
more lint

ryantxu committed:
check for nils

ryantxu committed:
check for nils

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
fix sql tests

ryantxu committed:
cleanup

ryantxu committed:
add missing file

ryantxu committed:
lint

ryantxu committed:
remove line

ryantxu committed:
remove line

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
setup test properly

ryantxu committed:
setup test properly

ryantxu commented on pkg/setting/setting_feature_toggles_test.go:
```diff
@@ -14,7 +14,6 @@ func TestFeatureToggles(t *testing.T) {
 		conf            map[string]string
 		err             error
 		expectedToggles map[string]bool
-		defaultToggles  map[string]bool
```
@bergquist -- this behavior is now defined by the initial expr in `registry.go`

ryantxu committed:
merge main

ryantxu committed:
remove more stuff

ryantxu committed:
remove more stuff

ryantxu committed:
remove more stuff

ryantxu commented on pkg/server/wire.go:
```diff
@@ -180,6 +181,8 @@ var wireBasicSet = wire.NewSet(
 	wire.Bind(new(teamguardian.Store), new(*teamguardianDatabase.TeamGuardianStoreImpl)),
 	teamguardianManager.ProvideService,
 	wire.Bind(new(teamguardian.TeamGuardian), new(*teamguardianManager.Service)),
+	featuremgmt.ProvideManagerService,
+	featuremgmt.ProvideToggles,
```
Perhaps over complicating this....  

Manager = can control (and eventually change) that flags
Toggles = readonly settings

DanCech replied:
It's a reasonable distinction, I had been considering a follow-up PR for licensing to do something similar as the majority of consumers only need a very small subset of the current interface.

ryantxu committed:
remove more stuff

ryantxu committed:
merge main

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
merge main

ryantxu committed:
fix sql tests

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
fix test

FZambia commented on pkg/services/featuremgmt/manager_test.go:
```diff
@@ -0,0 +1,31 @@
+
+func TestFeatureToggleSetup(t *testing.T) {
+	ft := WithFeatures("a", "b", "c")
+	assert.True(t, ft.IsEnabled("a"))
```
I prefer using `github.com/stretchr/testify/require` instead of `assert` since in most cases first error is what we aiming for when looking at test failure, the following errors in test are usually similar or caused by the first one etc - i.e. not really deterministic after first one.

gabor committed:
prometheus: fix auto-inserting null-timestamps at the end (#44022)

FZambia commented on pkg/services/featuremgmt/toggles_gen_test.go:
```diff
@@ -0,0 +1,144 @@
+
+func generateRegistry() (string, error) {
+	tmpl, err := template.New("fn").Parse(`
+// {{.CamleCase}} checks for the flag: {{.Flag.Name}}{{.Ext}}
```
```suggestion
// Is{{.CamleCase}}Enabled() checks for the flag: {{.Flag.Name}}{{.Ext}}
```

FZambia commented on pkg/services/featuremgmt/toggles_gen_test.go:
```diff
@@ -0,0 +1,144 @@
+	}
+
+	data := struct {
+		CamleCase string
```
Typo in word CamleCase

ryantxu replied:
:facepalm:

FZambia commented on pkg/services/featuremgmt/registry.go:
```diff
@@ -0,0 +1,162 @@
+			State:       FeatureStateStable,
+		},
+		{
+			Name:        "dashboardPreviews",
```
Can we somehow force the style for all next feature toggles, for example use camelCase only 🤔 Maybe at least write the suggested name format in FeatureFlag.Name field?

ryantxu replied:
In my original draft, I tried to fix that all up... but decided to punt until future PRs :)

FZambia commented on pkg/services/featuremgmt/features.go:
```diff
@@ -0,0 +1,95 @@
+	// FeatureStateBeta the feature is still in development, but settings will have migrations
+	FeatureStateBeta
+
+	// FeatureStateStable this is a stable feature
```
What FeatureStateStable means btw? Does stable state means that this should be enabled by default or moved to its own config section from feature toggles list? Or it can stay here forever? It seems to me all feature toggles should be removed eventually.

ryantxu replied:
TBD -- but I think we can/should potentially use feature flags for things that are in development *and* releasable.  If this is all statically defined on startup it does not make much difference, but as we add support for the flag to be set for some users and not others, that seems like something we want to be applied more generally.

I am also nervous that we have many parallel growing lists of features that can be enabled, and hope this more robust structure can give us a powerful single place to improve support

Jguer committed:
Add usage stats preview endpoint (#43899)

* Stats: add preview route for usage statistics

* Stats: respect reporting settings

* Stats: add tests to api endpoint

* Stats: unregister route

* Stats: always possible to preview stats

* Update pkg/infra/usagestats/service/api.go

Co-authored-by: Gabriel MABILLE <gamab@users.noreply.github.com>

Co-authored-by: Gabriel MABILLE <gamab@users.noreply.github.com>

FZambia commented on pkg/services/featuremgmt/manager.go:
```diff
@@ -0,0 +1,196 @@
+	return enabled
+}
+
+// IsEnabled checks if a feature is enabled
```
```suggestion
// Toggles returns FeatureToggles.
```

gamab committed:
AccessControl: Provisioning for teams (#43767)

* AccessControl: Provisioning for teams

* Docs: Add fixed role team assignment to full example

Co-authored-by: gamab <gabi.mabs@gmail.com>

Co-authored-by: Jguer <joao.guerreiro@grafana.com>

FZambia commented on pkg/services/featuremgmt/manager.go:
```diff
@@ -0,0 +1,196 @@
+		}
+
+		// Selectively update properties
+		if add.Description != flag.Description {
```
Should the check that add.Description is not empty be here? Moreover, maybe Description should be `*string` to allow overrides like this?

FZambia commented on pkg/services/featuremgmt/service.go:
```diff
@@ -0,0 +1,77 @@
+	return mgmt, nil
+}
+
+// interface without control
```
🤔

hugohaggmark committed:
Variables: Handle variable cancellations better (#43987)

Co-authored-by: Ashley Harrison <ashley.harrison@grafana.com>
Co-authored-by: joshhunt <josh@trtr.co>
Co-authored-by: kay delaney <kay@grafana.com>
Co-authored-by: Alexandra Vargas <alexa1866@gmail.com>>

marefr commented on pkg/services/featuremgmt/toggles_gen_test.go:
```diff
@@ -0,0 +1,144 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestFeatureToggleTypeScript(t *testing.T) {
```
Nit. A bit weird to do this in a test. Should probably use go generate.

ryantxu replied:
my go generate foo is not strong... I see how to get it to create functions from static files, but no idea (and can not find any examples) where it generates typescript and go based on a staticly defined struct.

I added it as a test because we want to fail if things ever get out of sync

marefr replied:
Okay :smile: We can look into that later if needed. There's probably many examples out on the www. It's just a lite weird that you need to run tests to generate the code and you want to commit the changes. In case you forget that you have to wait for CI to tell you about it

wbrowne committed:
[WIP] Plugins: Refactoring backend initialization flow (#42247)

* refactoring store interface and init flow

* fix import

* fix linter

* refactor resource calling

* load with class

* re-order args

* fix tests

* fix linter

* remove old creator

* add custom config struct

* fix some tests

* cleanup

* fix linter

* add connect failure error

* remove unused err

* convert test over

dsotirakis committed:
Update grabpl version to 2.8.4 (#44045)

fridgepoet committed:
Cloudwatch: Improve accessibility in metrics search and query builders (#43997)

* Add more aria labels on metrics search and query builders

* Use id for Input

Elfo404 committed:
Elasticsearch: Fix handling multiple datalinks for a single field (#44029)

JokerQueue committed:
Explore: Support custom display label for exemplar links for Prometheus datasource (#42732)

* Add custom URL label, docs for prometheus exemplar

Elfo404 committed:
Elasticsearch: Fix QueryEditor styling issues (#44041)

ToTheMax committed:
Fixed broken links in documentation (#44044)

* Fixed broken link

* Fixed broken contribute documentation link

* Consistent relative contribute-folder url

bergquist commented on packages/grafana-data/src/types/featureToggles.gen.ts:
```diff
@@ -1,14 +1,18 @@
  */
 export interface FeatureToggles {
-  [name: string]: boolean | undefined; // support any string value
+  // [name: string]?: boolean; // support any string value
```
```suggestion
```
?

ryantxu replied:
Ahh -- i just added a comment that we should uncomment this line before merging.  When it is commented out we get type errors when new frontend keys are created (3 in the last week!)

While developing it is nice to have in strict mode, but for release it should allow any string.

ryantxu committed:
Update pkg/services/featuremgmt/toggles_gen_test.go

Co-authored-by: Alexander Emelin <frvzmb@gmail.com>

ryantxu committed:
Update pkg/services/featuremgmt/manager.go

Co-authored-by: Alexander Emelin <frvzmb@gmail.com>

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
feature flags

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
update generated files

ryantxu committed:
keep the open structure

ryantxu commented on pkg/services/featuremgmt/features.go:
```diff
@@ -0,0 +1,95 @@
+	DocsURL     string             `json:"docsURL,omitempty"`
+
+	// CEL-GO expression.  Using the value "true" will mean this is on by default
+	Expression string `json:"expression,omitempty"`
```
this is really a placeholder for more complicated work from @bergquist and @oscarkilhed 's hackathon project.  See:
https://github.com/grafana/grafana/blob/experimentals/feature-toggle-by-headers/pkg/experiments/service.go#L141

ryantxu committed:
fix gosec

ryantxu committed:
typo

renovate[bot] committed:
Update dependency marked to v4.0.10 [SECURITY] (#44078)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

bergquist commented on pkg/services/featuremgmt/service.go:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	// Load config settings
+	configfile := filepath.Join(cfg.HomePath, "config", "features.yaml")
```
```suggestion
	configfile := filepath.Join(cfg.HomePath, "conf", "features.yaml")
```
For this to work we also need to create the file and include it in debian packages etc etc.

bergquist reviewed (commented):
A few nits. 

I think this is a step it the right direction and after some minor changes we can merge it. But I think we should consider the `features.yaml` feature unstable for a while. I want to have a bit more consensus about how we can operate this on scale first :)

eleijonmarck committed:
small fix :pray: (#44089)

kalleep committed:
Access Control: Allow signed in users access to GET data sources endpoints (#43338)

* remove scopes from endpoints and add datasources:read without scope to
the compatibility role

wbrowne committed:
fix delete plugin dashboard (#44055)

ashharrison90 committed:
Export: Fix error being thrown when exporting dashboards using query variables that reference the default datasource (#44034)

* handle undefined datasources, add unit test

* slightly nicer

mckn reviewed (commented):
Frontend code looks good! Great work with this @ryantxu

ashharrison90 committed:
Chore(CodeQL): Add noopener noreferrer to external links in email templates (#44092)

* Add noopener noreferrer to external links in email templates

* missed one...

* kick drone

hugohaggmark committed:
Chore: reduces circular dependencies for variables/utils.ts (#44087)

* Chore: move constants to own file

* Chore: moves safe* functions to grafana/data

* Chore: moves safe* functions to grafana/data

* Chore: adds VariableQueryEditorProps and deprecates VariableQueryProps

* Chore: remove getDefaultVariableAdapters function

* Chore: moves transaction status to types

* Chore: fix tests that do not initialise TemplateSrv

* Chore: change space when stringifying

* Chore: revert safe* func move

* Chore: remove circular dependency in Explore utils

gassiss committed:
Dashboard save interaction evt (#43304)

gillesdemey committed:
Alerting: show deleted datasource (#43891)

ivanahuckova committed:
Prometheus: Fix interpolation of $__rate_interval variable (#44035)

* Name fields based on targets length

* Remove rounding in rate interval variable replacing

* Refactor and fix edge  case when interval and time duration would be rate interval

* Add tests

* Remove , from test

* Revert "Name fields based on targets length"

This reverts commit 5dcb34a765f17693752ff8295be7a619788bdf29.

ifrost committed:
Elastic: Allow using long/int as date field for alerts (#44027)

* Use integers for time range filter

Previously it was passed as a string which is automatically converted by Elastic to a number only if the field type is "date". For other types (e.g. "long") such conversion doesn't work. In theory "date" could be passed as a formatted string but we don't use it this way and always pass it as a number so it is safe to always pass numbers, not strings.

* Fix time_series_query_test

* Retrigger build

sunker committed:
CloudMonitoring: Fixes broken variable queries that use group bys (#43914)

sunker committed:
Azure Monitor: Improved error messages for variable queries (#43213)

jackw committed:
Plugins: Add notices to docs to prevent NPX commands from hanging (#44043)

* docs(plugins): add notice about npm 7 flag to prevent commands from hanging

* Apply suggestions from code review

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

alexanderzobnin committed:
Access control: Team role picker (#43418)

* Refactor: move fetching from role picker to parent component

* Make built in role props optional

* Initial team role picker

* Add role picker to the teams list

* Optimize fetching roles

* Add pagination for the teams page

* Fix tests

* Hide roles if access control not enabled

* Fix test snapshots

* Refactor: use useAsync() hook

* Refactor: simplify input component

* Move api calls to separate file

* Refactor: use useAsync() hook for user role picker

* Tweak role picker input width

* Fix pagination

* Update test snapshots

* Use loading state from useAsync() hook

* Fix roles label if no roles assigned

malcolmholmes committed:
Release: remove bump from prepare release action (#44111)

mindw committed:
Fix issue link (#42891)

malcolmholmes committed:
Split prepare-release (#44124)

joshhunt committed:
Chore: Mark up User Profile page for translation (#43874)

* Mark up User profile page for translation

* Extract new messages

* updated selectors

* update selectors

* wip TestProvider

* update tests

* fix field labels

* extract new messages

* don't store date objects in redux state

* don't store date objects in redux state

tolzhabayev committed:
chore(checks): skip draft PRs (#43519)

* chore(checks): skip draft PRs for PR checks action

ryantxu committed:
Update pkg/services/featuremgmt/service.go

Co-authored-by: Carl Bergquist <carl.bergquist@gmail.com>

ryantxu committed:
Update pkg/services/featuremgmt/features.go

Co-authored-by: Carl Bergquist <carl.bergquist@gmail.com>

ryantxu committed:
merge main

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
Merge remote-tracking branch 'origin/main' into feature-flags

ryantxu committed:
keep comment about only true values

ryantxu committed:
minor cleanup

ryantxu committed:
lint

ryantxu committed:
merge main

ryantxu commented:
@DanCech @bergquist @xlson

I think this is ready for a final review/approval.  The enterprise PR is passing locally, but I think this has to be merged first before CI will help verify :grimacing:

DanCech commented:
If you have an enterprise branch with the same name CI will run test there against this branch instead of main

DanCech commented on pkg/services/featuremgmt/manager.go:
```diff
@@ -0,0 +1,199 @@
+			val = spec[idx].(bool)
+			idx++
+		}
+		flags[key] = FeatureFlag{
+			Name: key,
+		}
+
```
`flags` doesn't appear to be used

ryantxu committed:
merge main

ryantxu committed:
unused map

bergquist reviewed (approved):
I think we should continue working on this but I think we can merge this and move forward in another PR. Keeping this PR up to date will just be a massive drag.

marefr reviewed (approved):
LGTM

ryantxu committed:
merge main

ryantxu committed:
fix test

DanCech reviewed (approved):
Looks good!
